File: cpp/8-StringToInteger/main.cpp
#include <string>
#include <limits>
#include <iostream>

using namespace std;

class Solution 
{
private:
    string::iterator skipWhitespace(string::iterator it, string::iterator end) const
    {
        while(it != end && *it == ' ')
        {
            ++it;
        }

        return it;
    }

    bool isDigit(char c) const
    {
        return c >= '0' && c <= '9';
    }

    int toNumber(char c) const
    {
        return c - '0';
    }

public:
    int myAtoi(string s) 
    {
        auto end = s.end();
        auto it = skipWhitespace(s.begin(), end);

        if(it == end) return 0;

        bool isNegative = false;
        if(*it == '-' || *it == '+')
        {
            isNegative = (*it == '-');
            ++it;
        }

        uint64_t number = 0;
        uint64_t largestAbs = ((uint64_t)numeric_limits<int>::max()) + (isNegative ? 1 : 0);

        for(; it != end && isDigit(*it); ++it)
        {
            number = (number * 10) + toNumber(*it);

            if(number >= largestAbs) break;
        }

        if(isNegative && number >= largestAbs) return numeric_limits<int>::min();
        if(!isNegative && number >= largestAbs) return numeric_limits<int>::max();

        if(isNegative) number *= -1;

        return static_cast<int>(number);
    }
};

int main()
{
    Solution s;

    cout << s.myAtoi("1") << endl;
    cout << s.myAtoi("42") << endl;
    cout << s.myAtoi("   -42") << endl;
    cout << s.myAtoi("4193 with words") << endl;

    cout << s.myAtoi("2147483647") << endl;
    cout << s.myAtoi("2147483648") << endl;
    
    cout << s.myAtoi("-2147483648") << endl;

    return 0;
}

File: go/8-StringToInteger/main.go
package main

import (
	"fmt"
	"math"
)

func isDigit(c rune) bool {
	return c >= '0' && c <= '9'
}

func toNumber(c rune) int {
	return int(c - '0')
}

func skipWhitespace(s []rune) []rune {
	for i, c := range s {
		if c != ' ' {
			return s[i:]
		}
	}

	return nil
}

func myAtoi(s string) int {
	runes := skipWhitespace([]rune(s))

	if len(runes) == 0 {
		return 0
	}

	isNegative := false
	if runes[0] == '-' || runes[0] == '+' {
		isNegative = runes[0] == '-'
		runes = runes[1:]
	}

	var number int64 = 0

	for i := 0; i < len(runes) && isDigit(runes[i]); i++ {
		number = (number * 10) + int64(toNumber(runes[i]))

		if number > math.MaxInt32 {
			break
		}
	}

	if isNegative {
		number *= -1
	}

	if number > math.MaxInt32 {
		number = math.MaxInt32
	} else if number < math.MinInt32 {
		number = math.MinInt32
	}

	return int(number)
}

func test1() {
	fmt.Println(myAtoi("42"))
}

func test2() {
	fmt.Println(myAtoi("-042"))
}

func test3() {
	fmt.Println(myAtoi("1337c0d3"))
}

func test4() {
	fmt.Println(myAtoi("0-1"))
}

func test5() {
	fmt.Println(myAtoi("words and 987"))
}

func test6() {
	fmt.Println(myAtoi("2147483647"))
}

func test7() {
	fmt.Println(myAtoi("2147483649"))
}

func main() {
	test1()
	test2()
	test3()
	test4()
	test5()
	test6()
	test7()
}
